### PR TITLE
Add watch-only mode

### LINKED ISSUE #238: Add watch-only mode
## Goal
Zingolib can be used to scan blockchain using only viewing keys.

## Motivation
In SatoshiLabs, we are working on shielded transactions for Trezor T hardware wallet (HWW). Since fast scanning of Zcash blockchain is a pretty complex task, we would like to delegate it to an existing library. We cannot use zingolib as it is, because at the moment the zingo-client must be initiated with a spending key, which never leaves a HWW.

## History
Crate [zecwallet-light-cli](https://github.com/adityapk00/zecwallet-light-cli) (ancestor of zingolib) was able to import viewing keys via `import` command. During a blockchain scanning, zecwallet was using all available keys and it was labeling obtained notes to be spendable and watch-only.

We don't want to come back to this design because
- one-key-per-poll architecture is simpler
- scanning blockchain with multiple keys would be slow

## Proposed solution
At the moment, zingolib uses [`UnifiedSpendCapability`](https://github.com/zingolabs/zingolib/blob/dev/zingolib/src/wallet/keys/unified.rs#L19-L32) to store spending keys.

I want to add more types of "capabilities"
  ```rust
pub enum Capability<SpendKeyType, ViewingKeyType> {
	Spend(SpendKeyType),
	View(ViewingKeyType),
	None,
}
```
and then replace the `UnifiedSpendAuthority` by `WalletCapability`
```rust
pub struct WalletCapability {
	transparent: Capability<...>,
	sapling: Capability<...>,
	orchard: Capability<...>,
	addresses: Vec<UnifiedAddress>,
	next_sapling_diversifier_index: DiversifierIndex,
}
```

#### Why is there a `Capability::None` option?
It allows wallet to be friendly to unified viewing keys that do not contain a viewing key for every pool.

Further, the `Capability::None` option is useful to reflect this fact, that HWW does not support some pool. E.g. Trezor probably will not support Sapling and Ledger has no support for Orchard.

#### Why not add something like `UnifiedViewingCapability`?
It seems to be more idiomatic in Rust to pass one enum representing all possible options, then having two structs with embedded `Option<...>` (for `None`-capability mentioned above) and converting between them, while catching errors...

#### How would this design change, if direct communication with HWW is implemented into zingolib?
  ```rust
pub enum Capability<SpendKeyType, ViewingKeyType> {
	Spend(SpendKeyType),
	SpendViaHWW(ViewingKeyType),
	View(ViewingKeyType),
	None,
}
```

### PR DESCRIPTION
**Pull request by jarys**
closes #238

### Changes
Note: by being optional I mean replacing type `T` with type `Option<T>`.

- `UnifiedSpendAuthority` replaced with `WalletCapability`
    - See #238 for more details.
- `WalletCapability as ReadableWriteable<()>` version incremented to 2, while staying compatible with version 1 `UnifiedSpendCapability as ReadableWriteable<()>` 
- `LightClient::create_with_seedorkey_wallet` renamed to `LightClient::new_from_wallet_base`
    - and its argument `seed_or_key: Option<String>` replaced with `wallet_base: WalletBase`
```rust
pub enum WalletBase {
    FreshEntropy,
    SeedBytes([u8; 32]),
    MnemonicPhrase(String),
    Mnemonic(Mnemonic),
    Ufvk(String),
}
```
- `UnifiedSpendAuthority: Into<SomeKey>` replaced with `WalletCapability: TryInto<SomeKey>`
    - i.e. all these conversions are `WalletCapability` failable
- arguments `orchard_ivk` and `sapling_ivk` of `TrialDecryptions::trial_decrypt_batch` are optional now
- field `mnemonic` of `LightWallet` is now optional
- `LightWallet::serialized_version()` incremented to 27 to enable optional mnemonic serialization
- `LightWallet::send_to_address_inner` returns error in watch-only mode
- field `extsk` of `SpendableSaplingNote` and field `spend_key` of `SpendableOrchardNote` are now optional
- added `ExtendedPubKey` struct as complement of `ExtendedPrivKey`
- added `secp256k1::SecretKey: ReadableWriteable<()>`
- added `OrchardSpendingKey: ReadableWriteable<()>`
- added `SaplingExtendedSpendingKey: ReadableWriteable<()>`
- added `--viewing-key` argument for `zingo-cli` :sparkles:

### CODE REVIEW

**Comment by jarys** [comment]
PR is ready for review @AloeareV

commits should be be squashed before merging

there are no test scenarios at the moment . I will add some , this or next week . I tested on my own and everything was working as expected :sparkles: 

cc: @zancas

File: zingolib/src/wallet/keys/unified.rs
Comment (zancas) [inline zingolib/src/wallet/keys/unified.rs:140]: "Wallet is not capable of producing desired receivers."

File: zingolib/src/wallet/keys/unified.rs
Comment (zancas) [inline zingolib/src/wallet/keys/unified.rs:27]: @AloeareV proposes making this `#[non_exhaustive']` both the future proof against `SpendViaHWW(ViewingKeyType)`, and to future proof against allowing different types of viewing keys `Outgoing` `Incoming`.

File: zingolib/src/wallet/keys/unified.rs
Comment (zancas) [inline zingolib/src/wallet/keys/unified.rs:128]: "The wallet is not capable of spending transparent funds."

**Review by AloeareV — CHANGES_REQUESTED** [review]
Just a couple notes, this looks great! Needs tests, but I'd be happy to include those in a subsequent PR in order to get this in before more merge conflicts crop up.

File: zingolib/src/lightclient.rs
Comment (zancas) [inline zingolib/src/lightclient.rs:504]: Nice!

**Review by zancas — COMMENTED** [review]
This looks awesome.   We definitely need to add some tests.

**Review by zancas — CHANGES_REQUESTED** [review]
We need to add tests that show proper handling of new-and-old wallet serialization formats.

File: zingolib/src/lightclient.rs
Comment (jarys) [reply zingolib/src/lightclient.rs:504]: 👌

File: zingolib/src/wallet/keys/unified.rs
Comment (jarys) [reply zingolib/src/wallet/keys/unified.rs]: 3803f14f23f0bbf7229ca7878ffbdd86fba887db

File: zingolib/src/wallet/keys/unified.rs
Comment (jarys) [reply zingolib/src/wallet/keys/unified.rs]: a84c39daf77360b612a7bd532058753248b77793

File: zingolib/src/wallet/keys/unified.rs
Comment (jarys) [reply zingolib/src/wallet/keys/unified.rs:28]: 3a73b7f11c9338ac302895471672502f197eca96

**Comment by jarys** [comment]
Sure, making `enum Capability` `#[non_exhaustive]` is definitely appropriate. I didn't know about this attribute.

**Comment by jarys** [comment]
I will focus on tests now.

**Comment by zancas** [comment]
One other ask I have is to make a type, for the Error variant of Results returned from public `fn`s.

**Review by AloeareV — DISMISSED** [review]

**Review by zancas — DISMISSED** [review]

**Comment by jarys** [comment]
commits squashed . this is [no-op](https://github.com/zingolabs/zingolib/compare/7eab4a7..07bc566)

**Comment by jarys** [comment]
conflicts resolved . this PR is ready to be merged

**Comment by jarys** [comment]
I do force push one more time to rebase instead of merge.

**Comment by jarys** [comment]
I found a mistake when I was writing tests. I was trying to load Sapling `ExtendedFullViewingKey`, but only `DiversifiableFullViewingKey` is included in UFVK.

**Comment by jarys** [comment]
So I replaced Sapling Extended FVK with Diversifiable FVK (6cd0f4946e812a9a8cf455c9181c0e3ce9bb5831)

I removed `TransactionMetadatSet::adjust_spendable_status` (49cd06f4a20a8f8c2de9f3affa03052a742ebaad). This method requires `Fvk: PartialEq`, which is not implemented for Diversifiable FVK. But this method is outdated anyway. It was used in version 8 and below [[source](https://github.com/adityapk00/zecwallet-light-cli/blob/fd5d11f0f28e0f1628ea8fdad0cce16d50e6bb98/lib/src/lightwallet.rs#L383-L393)]

And I fixed `scenario_mempool_clearing` test (b71f4c42bbcd7907820792bbd1ae67007781e747). This test started failing, because it was using Debug print to compare curve points represented by non-normalized projective coordinates 🤷 yikwim

**Comment by jarys** [comment]
Tests are finished. Blocked by
- #256

**Review by AloeareV — APPROVED** [review]

**Review by AloeareV — APPROVED** [review]

**Comment by AloeareV** [comment]
@zancas @juanky201271 Note that this PR makes a change to the zingolib public interface that zingo-mobile uses, renames `LightClient::create_with_seedorkey_wallet` to `LightClient::new_from_wallet_base`, and it now takes a `WalletBase` instead of a `String` as an argument.

**Pull request merged**
